File: 0004-median-of-two-sorted-arrays/0004-median-of-two-sorted-arrays.go
func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
    if len(nums2) < len(nums1) {
        nums1, nums2 = nums2, nums1
    }
    fmt.Println(nums1)
    len1, len2 := len(nums1), len(nums2)
    full := len1 + len2
    
    low, high := 0, len1
    for low <= high {
        mid1 := (low + high) / 2
        mid2 := full / 2 - mid1
        
        left1, right1 := math.MinInt32, math.MaxInt32
        if 0 < mid1 {
            left1 = nums1[mid1-1]
        }
        if mid1 < len1 {
            right1 = nums1[mid1]
        }
        
        left2, right2 := math.MinInt32, math.MaxInt32
        if 0 < mid2 {
            left2 = nums2[mid2-1]
        }
        if mid2 < len2 {
            right2 = nums2[mid2]
        }
        
        if med, ok := median(left1, right1, left2, right2, full); ok {
            return med
        } else if right2 < left1 {
            high = mid1 - 1
        } else {
            low = mid1 + 1
        }
    }
    
    return 0
}

func median(l1, r1, l2, r2, full int) (float64, bool) {
    if l1 <= r2 && l2 <= r1 {
        if full % 2 == 0 {
            return float64((max(l1, l2) + min(r1, r2))) / 2, true
        }
        return float64(min(r1, r2)), true
    }
    return 0, false
}

func max(a, b int) int {
    if a < b {
        return b
    }
    return a
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

File: 0004-median-of-two-sorted-arrays/0004-median-of-two-sorted-arrays.py
INT_MIN = -(2 ** 32)
INT_MAX = 2 ** 32

class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        n1 = len(nums1)
        n2 = len(nums2)
        
        if n2 < n1:
            nums1, nums2 = nums2, nums1
            n1, n2 = n2, n1
        
        low, high = 0, n1
        while low <= high:
            mid1 = (low + high) // 2
            mid2 = (n1 + n2) // 2 - mid1

            l1, r1 = INT_MIN, INT_MAX
            if 0 < mid1:
                l1 = nums1[mid1-1]
            if mid1 < n1:
                r1 = nums1[mid1]
            
            l2, r2 = INT_MIN, INT_MAX
            if 0 < mid2:
                l2 = nums2[mid2-1]
            if mid2 < n2:
                r2 = nums2[mid2]
            
            if l1 <= r2 and l2 <= r1:
                if (n1 + n2) % 2 == 0:
                    return (max(l1, l2) + min(r1, r2)) / 2
                else:
                    return min(r1, r2)
            elif r2 < l1:
                high = mid1 - 1
            else:
                low = mid1 + 1
                
        return 0.0        